htop.1.in
.TH "HTOP" "1" "2026" "@PACKAGE_STRING@" "Utils"
.SH "NAME"
htop \- interactive process viewer
.SH "SYNOPSIS"
.LP 
.B htop [\fIOPTIONS\fR]
.SH "DESCRIPTION"
.LP 
Htop is a free (GPL) ncurses-based process viewer and system monitor for
Unix-like systems.
.LP
It is similar to top, but allows you to scroll vertically and horizontally,
so you can see all the processes running on the system, along with their full
command lines, as well as viewing them as a process tree, selecting multiple
processes and acting on them all at once.
.LP
Tasks related to processes (killing, renicing) can be done without
entering their PIDs.
.br 
.SH "COMMAND-LINE OPTIONS"
.LP
Mandatory arguments to long options are mandatory for short options too.
.LP 
.TP
\fB\-d, \-\-delay \fIDELAY\fR
Delay between updates, in tenths of seconds. If the delay value is
less than 1 it is increased to 1, i.e. 1/10 second. If the delay value
is greater than 100, it is decreased to 100, i.e. 10 seconds.
.TP
\fB\-C, \-\-no-color, \-\-no-colour\fR
Start htop in monochrome mode
.TP
\fB\-h, \-\-help
Display a help message and exit
.TP
\fB\-p, \-\-pid \fIPID\fR[\fB,\fIPID...\fR]
Show only the given PIDs
.TP
\fB\-s, \-\-sort\-key \fICOLUMN\fR
Sort by this column (use \-\-sort\-key help for a column list)
.TP
\fB\-u, \-\-user \fIUSERNAME\fR
Show only the processes of a given user
.TP
\fB\-v, \-\-version
Output version information and exit
.TP
\fB\-t, \-\-tree
Show processes in tree view
@disk_stats_man_page_paragraph@
.TP
\fB\-\-explicit-delay
Explicitly delay between updates; this could be useful to workaround ncurses
timer issues when the system time is changed
.PP
.br 
.SH "INTERACTIVE COMMANDS"
.LP 
The following commands are supported while in htop:
.LP 
.TP 5
.B Up, Alt-k
Select (highlight) the previous process in the process list. Scroll the list
if necessary.
.TP
.B Down, Alt-j
Select (highlight) the next process in the process list. Scroll the list if
necessary.
.TP
.B Left, Alt-h
Scroll the process list left.
.TP
.B Right, Alt-l
Scroll the process list right.
.TP
.B PgUp, PgDn
Scroll the process list up or down one window.
.TP
.B Home
Scroll to the top of the process list and select the first process.
.TP
.B End
Scroll to the bottom of the process list and select the last process.
.TP
.B Ctrl-A, ^
Scroll left to the beginning of the process entry (i.e. beginning of line).
.TP
.B Ctrl-E, $
Scroll right to the end of the process entry (i.e. end of line).
.TP
.B Space
Tag or untag a process. Commands that can operate on multiple processes,
like "kill", will then apply over the list of tagged processes, instead
of the currently highlighted one.
.TP
.B c
Tag the current process and its children. Commands that can operate on multiple
processes, like "kill", will then apply over the list of tagged processes,
instead of the currently highlighted one.
.TP
.B U
Untag all processes (remove all tags added with the Space or c keys).
.TP
.B s
Trace process system calls: if truss(1) or strace(1) is installed, pressing
this key will attach it to the currently selected process, presenting a live
update of system calls issued by the process.
.TP
.B l, o
Display open files for a process: if lsof(1) is installed, pressing this key
will display the list of file descriptors opened by the process.
.TP
.B A
Display full command line arguments of the selected process in a separated
screen.
.TP
.B e
Display environment variables of the selected process in a separated screen.
.TP
.B S
Display kernel stack trace of the selected process in a separated screen.
.TP
.B v
Display virtual memory mappings of the selected process in a separated screen.
.TP
.B F1, h, ?
Go to the help screen
.TP
.B F2, C
Go to the setup screen, where you can configure the meters displayed at the top
of the screen, set various display options, choose among color schemes, and
select which columns are displayed, in which order.
.TP
.B F3, /
Incrementally search the command lines of all the displayed processes. The
currently selected (highlighted) command will update as you type. While in
search mode, pressing F3 will cycle through matching occurrences.
.TP
.B n
Cycle through matching occurrences of last search.
.TP
.B N
Cycle through matching occurrences of last search in reverse order.
.TP
.B F4, \\\\
Incremental process filtering: type in part of a process command line and
only processes whose names match will be shown. To cancel filtering,
enter the Filter option again and press Esc.
.TP
.B F5, t
Tree view: organize processes by parenthood, and layout the relations
between them as a tree. Toggling the key will switch between tree and
your previously selected sort view. Selecting a sort view will exit
tree view.
.TP
.B F6
On sorted view, select a field for sorting, also accessible through < and >.
The current sort field is indicated by a highlight in the header.
On tree view, expand or collapse the current subtree. A "+" indicator in the
tree node indicates that it is collapsed.
.TP
.B F7, ]
Increase the selected process's priority (subtract from 'nice' value).
This can only be done by the superuser.
.TP
.B F8, [
Decrease the selected process's priority (add to 'nice' value)
.TP
.B F9, k
"Kill" process: sends a signal which is selected in a menu, to one or a group
of processes. If processes were tagged, sends the signal to all tagged processes.
If none is tagged, sends to the currently selected process.
.TP
.B F10, q
Quit
.TP
.B I
Invert the sort order: if sort order is increasing, switch to decreasing, and
vice-versa.
.TP
.B +, \- 
When in tree view mode, expand or collapse subtree. When a subtree is collapsed
a "+" sign shows to the left of the process name.
.TP
.B a (on multiprocessor machines)
Set CPU affinity: mark which CPUs a process is allowed to use.
.TP
.B u
Show only processes owned by a specified user.
.TP
.B M
Sort by memory usage (top compatibility key).
.TP
.B P
Sort by processor usage (top compatibility key).
.TP
.B T
Sort by time (top compatibility key).
.TP
.B F
"Follow" process: if the sort order causes the currently selected process
to move in the list, make the selection bar follow it. This is useful for
monitoring a process: this way, you can keep a process always visible on
screen. When a movement key is used, "follow" loses effect.
.TP
.B K
Hide kernel processes: prevent the kernel processes to be displayed in the
process list. (This is a toggle key.)
.TP
.B H
Hide user threads: on systems that represent them differently than ordinary
processes (such as recent NPTL-based systems), this can hide threads from
userspace processes in the process list. (This is a toggle key.)
.TP
.B p
Show full paths to running programs, where applicable. (This is a toggle key.)
.TP
.B Ctrl-L
Refresh: redraw screen and recalculate values.
.TP
.B Numbers
PID search: type in process ID and the selection highlight will be moved to it.
.PD

.LP
When the \fIvi(1)-style key-bindings\fR setting is enabled, some key bindings
are redefined as followings:
.LP 
.TP 5
.B Numbers
If issuing the next command repeatly is meaningful, set the issuing count
for the key immediately after the numbers. The minimal count is \fI1\fR.
.TP
.B k, j, h, l
Same as \fBUp\fR, \fBDown\fR, \fBLeft\fR and \fBRight\fR respectively.
.TP
.B Alt-k, Alt-j, Alt-h, Alt-l
Not available under this mode.
.PD

.SH "COLUMNS"
.LP 
The following columns can display data about each process. A value of '\-' in
all the rows indicates that a column is unsupported on your system, or
currently unimplemented in htop. The names below are the ones used in the
"Available Columns" section of the setup screen. If a different name is
shown in htop's main screen, it is shown below in parenthesis.
.LP 
.TP 5
.B Command
The full command line of the process (i.e. program name and arguments).
.TP 
.B PID
The process ID.
.TP
.B NAME
Process name, usually the executable file name.
.TP
.B STATE (S)
The state of the process:
   \fBS\fR for sleeping
   \fBI\fR for sleeping longer than an implementation-defined time
   \fBR\fR for running or runnable
   \fBO\fR for running on a processor
   \fBD\fR for uninterruptible waiting, usually for disk I/O
   \fBZ\fR for zombie (waiting for parent to read its exit status)
   \fBT\fR for stopped due to signal or being traced
   \fBL\fR for waiting on a lock
   \fBW\fR for an implementation-defined state (e.g. paging for Linux)

Not all states are supported on every platform.
.TP
.B PPID
The parent process ID.
.TP
.B PGRP
The process's group ID.
.TP
.B SESSION (SID)
The process's session ID.
.TP 
.B TTY
The controlling terminal of the process.
.TP
.B TPGID
The process ID of the foreground process group of the controlling terminal.
.TP
.B MINFLT
The number of page faults happening in the main memory.
.TP
.B CMINFLT
The number of minor faults for the process's waited-for children (see MINFLT above).
.TP
.B MAJFLT
The number of page faults happening out of the main memory.
.TP
.B CMAJFLT
The number of major faults for the process's waited-for children (see MAJFLT above).
.TP
.B UTIME (UTIME+)
The user CPU time, which is the amount of time the process has spent executing
on the CPU in user mode (i.e. everything but system calls), measured in clock
ticks.
.TP
.B STIME (STIME+)
The system CPU time, which is the amount of time the kernel has spent
executing system calls on behalf of the process, measured in clock ticks.
.TP
.B CUTIME (CUTIME+)
The children's user CPU time, which is the amount of time the process's
waited-for children have spent executing in user mode (see UTIME above).
.TP
.B CSTIME (CSTIME+)
The children's system CPU time, which is the amount of time the kernel has spent
executing system calls on behalf of all the process's waited-for children (see
STIME above).
.TP
.B PRIORITY (PRI)
The kernel's internal priority for the process, usually just its nice value
plus twenty. Different for real-time processes.
.TP
.B NICE (NI)
The nice value of a process, from 19 (low priority) to -20 (high priority). A
high value means the process is being nice, letting others have a higher
relative priority. The usual OS permission restrictions for adjusting priority apply.
.TP
.B STARTTIME (START)
The time the process was started.
.TP
.B PROCESSOR (CPU)
The ID of the CPU the process last executed on.
.TP
.B M_SIZE (VIRT)
The size of the virtual memory of the process.
.TP
.B M_RESIDENT (RES)
The resident set size (text + data + stack) of the process (i.e. the size of the
process's used physical memory).
.TP
.B M_SHARE (SHR)
The size of the process's shared pages.
.TP
.B M_TRS (CODE)
The text resident set size of the process (i.e. the size of the process's
executable instructions).
.TP
.B M_DRS (DATA)
The data resident set size (data + stack) of the process (i.e. the size of anything
except the process's executable instructions).
.TP
.B M_LRS (LIB)
The library size of the process.
.TP
.B M_DT (DIRTY)
The size of the dirty pages of the process.
.TP
.B EFFECTIVE_UID (EUID)
The effective user ID of the process.
.TP
.B REAL_UID (RUID)
The real user ID of the process.
.TP
.B EFFECTIVE_USER (EFFE_USER)
Effective user (or numeric user ID if name cannot be determined) of the process.
.TP
.B REAL_USER
Real user (or numeric user ID if name cannot be determined) of the process.
.TP
.B PERCENT_CPU (CPU%)
The percentage of the CPU time that the process is currently using.
.TP
.B PERCENT_MEM (MEM%)
The percentage of memory the process is currently using (based on the process's
resident memory size, see M_RESIDENT above).
.TP
.B TIME (TIME+)
The time, measured in clock ticks that the process has spent in user and system
time (see UTIME, STIME above).
.TP
.B NLWP
The number of threads in the process.
.TP
.B TGID
The thread group ID. Available on platforms where threads are presented as
processes.
.TP
.B CTID
OpenVZ container ID, a.k.a virtual environment ID.
.TP
.B VPID
OpenVZ process ID.
.TP
.B VXID
VServer process ID.
.TP
.B RCHAR (RD_CHAR)
The number of bytes the process has read.
.TP
.B WCHAR (WR_CHAR)
The number of bytes the process has written.
.TP
.B SYSCR (RD_SYSC)
The number of read(2) syscalls for the process.
.TP
.B SYSCW (WR_SYSC)
The number of write(2) syscalls for the process.
.TP
.B RBYTES (IO_RBYTES)
Bytes of read(2) I/O for the process.
.TP
.B WBYTES (IO_WBYTES)
Bytes of write(2) I/O for the process.
.TP
.B CNCLWB (IO_CANCEL)
Bytes of cancelled write(2) I/O.
.TP
.B IO_READ_RATE (DISK READ)
The I/O rate of read(2) in bytes per second, for the process.
.TP
.B IO_WRITE_RATE (DISK WRITE)
The I/O rate of write(2) in bytes per second, for the process.
.TP
.B IO_RATE (DISK R/W)
The I/O rate, IO_READ_RATE + IO_WRITE_RATE (see above).
.TP
.B CGROUP
Which cgroup the process is in.
.TP
.B OOM
OOM killer score.
.TP
.B IO_PRIORITY (IO)
The I/O scheduling class followed by the priority if the class supports it:
   \fBR\fR for Realtime
   \fBB\fR for Best-effort
   \fBid\fR for Idle
.TP
.B PERCENT_CPU_DELAY (CPUD%)
The percentage of time spent waiting for a CPU (while runnable). Requires CAP_NET_ADMIN.
.TP
.B PERCENT_IO_DELAY (IOD%)
The percentage of time spent waiting for the completion of synchronous block I/O. Requires CAP_NET_ADMIN.
.TP
.B PERCENT_SWAP_DELAY (SWAPD%)
The percentage of time spent swapping in pages. Requires CAP_NET_ADMIN.
.TP
.B All other flags
Currently unsupported (always displays '-').

.SH "CONFIG FILE"
.LP 
By default htop reads its configuration from the XDG-compliant path
\fI~/.config/htop/htoprc.$TERM\fR -- where \fI$TERM\fR is the terminal type
htop currently runs with; the configuration file is overwritten by htop's
in-program Setup screen, so it should not be hand-edited. If no
terminal-type-specific configuration file exists, htop then tries to read
\fI~/.config/htop/htoprc\fR; if that is also missing it tries the system-wide
configuration from \fI@sysconfdir@/htoprc\fR and as a last resort, before
falls back to its hard coded defaults.
.LP
You may override the location of the configuration file using the \fBHTOPRC\fR
environment variable (so you can have multiple configurations for different
machines that share the same home directory, for example).

.SH "USER-DEFINED COLOR SCHEMES"
.LP
In addition to built-in color schemes, htop supports using more color schemes
defined from external files. Htop tries to read files under directory
\fI~/.config/htop/colorschemes\fR as color scheme definition files;
successfully loaded color schemes will appear in the Setup screen. Please see
\fIhtop.colorscheme(5)\fR for the file format details.

.SH "MEMORY SIZES"
.LP
Memory sizes in htop are displayed as they are in tools from the GNU Coreutils
(when ran with the --human-readable option). This means that sizes are printed
in powers of 1024. (e.g., 1023M = 1072693248 Bytes)
.LP
The decision to use this convention was made in order to conserve screen space
and make memory size representations consistent throughout htop.

.SH "SEE ALSO"
proc(5), top(1), free(1), ps(1), uptime(1), kill(1), vmstat(8)

.SH "AUTHORS"
.LP 
htop is developed by Hisham Muhammad <hisham@gobolinux.org>.
.LP
This man page was written by Bartosz Fenski <fenio@o2.pl> for the Debian
GNU/Linux distribution (but it may be used by others). It was updated by Hisham
Muhammad, and later by Vincent Launchbury, who wrote the 'Columns' section.


ChangeLog
What's new in version 2.3.1

* Process open file count column for Interix
* BUGFIX: avoid memory size values underflow when reading from Linux procfs
* BUGFIX: reduce changes of memory size values overflow by always using a
  generally big enough integer type to store and calculate them
* Improve fatal error reporting on Darwin
* BUGFIX: fix process list sometimes stops sorting
* Disk statistics mode for Linux-based and kFreeBSD-based platforms
* Generic support for colorization of percent values
* BUGFIX: fix processor time of processes losing precision on OpenBSD
* BUGFIX: fix detection of multi-processor system fails when Linux sysfs
  '/sys/devices/system/cpu/' directory is empty
* BUGFIX: fix incorrect processor percentage of processes when a Linux-based
  system have any off-line processor
* Support of AIX versions 6.1 and 7.2
* BUGFIX: fix incorrect arguments to lseek(2) in kFreeBSD-specific code
* Mac OS X Tiger powerpc support
* Support using Mach task_info(2) to get additional task information without
  libproc for older releases of Mac OS X
* Check and make sure the last process field is HTOP_COMM_FIELD when the
  configuration is loaded from legacy or global configuration file; this helps
  transiting from configuration files generated in version 2.2.0 or earlier
* Support moving selected panel entry to top or buttom by pressing Home or End
  key respectively
* BUGFIX: fix Enter key sometimes didn't work as expected
* Build system: deprecate uses of 'ncurses*-config' scripts, they will no
  longer be used unless enabled explicitly via environment variable(s)
* Build system: try to use pkg-config(1) to get requires flags for ncurses,
  before falling back to previous detection methods
* Adjust text format of some meters to improve readability

What's new in version 2.3.0

* System call tracing support via truss(1)
* Process name column
* AIX support
  (contributed by Calvin Buckley)
* BUGFIX: correctly show command line of processes on Solaris
* BUGFIX: fix process time scaling error on Solaris
* BUGFIX: fix nice value of processes on Solaris
* Build system: translate / rewrite MakeHeaders.py into MakeHeaders.sh
* BUGFIX: fix defining variables in headers
* ZFS ARC size display in existing memory meter
* BUGFIX: fix file descriptor leaks in OpenFilesScreen
* BUGFIX: fix hard coded names for reading battery information from Linux
  sysfs
* Android/Linux support
* BUGFIX: null pointer dereference and memory corruption fixes for OpenBSD
* CPU hot-plugging support on Linux-based platforms and Solaris
* Remove dependency to the BSD-specific err(3)
* Distinguishable effective user and real user of processes
* Fix incorrect buffer sizes for snprintf(3) and strftime(3) calls; make more
  efficient use of stack space
* Basic Interix support
* Cygwin support
* Thread processes support for OpenBSD
* OpenBSD support down to release 5.0
* Optinally display kernel processes in a different color
* Solaris support without libproc(3LIB)
* BUGFIX: fix some sysctl(3) data type mismatches
* BUGFIX: enanbling hide kernel processes no longer hiding everything if the
  root process is a kernel process
* BUGFIX: properly update modifiable information for existing processes
* Colorize process state characters in help screen
* BUGFIX: fix pressing PPAGE or NPAGE when moving a selected meter or column
  selects another item instead of move the original item
* Change a tree-drawing character to make the process tree more readable in
  ASCII mode
* Try to enable backtrace(3) support for all platforms
* Detect and workaround curses(3) macro ERR being incorrectly redefined by
  other system headers (such as sys/regset.h on Solaris x86)
* BUGFIX: make function Header_addMeterByName no longer modifying the passed
  in string buffer
* Process environment support for DragonFly BSD, and kFreeBSD-based platforms
* kFreeBSD-based platforms support without kvm(3)
* FreeBSD support down to release 5
* BUGFIX: correctly detect kernel processes on Linux-based platforms
* Build-time detection of available signals
* Widen UID column to 6-character space for showing values
* Support pressing n or N to search next or previous in MainPanel and
  InfoScreen
  (thanks to syrrim for the initial implementation for MainPanel)
* BUGFIX: fix bad coding style and the caused error of using uninitialized
  data in function 'InfoScreen_run'
* BUGFIX: fix 'break' without a dedicated line in switch block caused
  incorrect 'break' statements
* Add ArgScreen to list command line arguments of a process
* BUGFIX: don't follow process when selecting non-process-specific options
  (thanks to wangqr)
* Processor times meter improvements for OpenBSD
  (thanks to Antoine Motet)
* BUGFIX: rename many enumeration constants to avoid identifier pollution
* BUGFIX: don't unfollow process on LEFT or RIGHT key
* BUGFIX: fix incorrect maximum PID value assumption for Solaris
* Honour setting of counting CPUs from 0/1 when presenting CPU meters
  (thanks to Ricardo Nabinger Sanchez)
* Exit on invalid command line usage
* vi(1)-style key bindings support
* Add danish translation to htop.desktop
  (thanks to scootergrisen)
* BUGFIX: always reformat process start time on display
* Mouse support can now be disabled in setup
* BUGFIX: handle circularized process tree(s) that may happening with some
  kernels
* BUGFIX: fix "Broken Gray" didn't change the color of cpu-iowait
  (thanks to fo40225)
* Support loading user-defined color schemes, with 2 user-defined color
  schemes added in source tree as examples
  (thanks to ajdiaz and Alan Jenkins for contributing the color schemes)
* BUGFIX: place cursor in appropriate location during incremental searching
  in MainPanel
* BUGFIX: consistent process state names
* Use actual null pointer instead of null pointer constant for execlp(3)
  (thanks to Kang-Che Sung for pointing out)
* BUGFIX: fix passing negative value other than -1 to ctype function leading
  to undefined behavior
* Add HTOP_PROCESS_Z_STATE_COLOR

htop.c
/*
htop - htop.c
(C) 2004-2011 Hisham H. Muhammad
Copyright 2015-2026 Rivoreo
Released under the GNU GPL, see the COPYING file
in the source distribution for its full text.
*/

#include "config.h"

#if defined DISK_STATS && !defined HAVE_GETOPT_LONG
#error "Disk statistics mode requires getopt_long(3)"
#endif

#include "FunctionBar.h"
#include "Hashtable.h"
#include "ColumnsPanel.h"
#include "CRT.h"
#include "MainPanel.h"
#include "ProcessList.h"
#include "ScreenManager.h"
#include "Settings.h"
#include "UsersTable.h"
#include "Platform.h"
#ifdef DISK_STATS
#include "DiskPanel.h"
#include "DiskList.h"
#include "Disk.h"
#endif
#if defined HAVE_GETOPT_H && defined HAVE_GETOPT_LONG
#include <getopt.h>
#endif
#include <locale.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

static void printVersionFlag() {
   fputs("htop " VERSION "\n" COPYRIGHT "\n"
         "Released under the GNU GPL any version.\n",
         stdout);
}

static void print_usage(FILE *f, const char *name) {
   fprintf(f,
         "Usage: %s [OPTION]...\n\n"
         "Options:\n"
#ifdef HAVE_GETOPT_LONG
         "   -C, --no-color              Use a monochrome color scheme\n"
         "   -d, --delay=DELAY           Set the delay between updates, in tenths of\n"
         "                               seconds\n"
         "   -h, --help                  Print this help screen\n"
         "   -s, --sort-key=COLUMN       Sort by COLUMN (try --sort-key=help for a list)\n"
         "   -t, --tree                  Show the tree view by default\n"
         "   -u, --user=USERNAME         Show only processes of a given user\n"
         "   -p, --pid=PID[,PID,PID...]  Show only the given PIDs\n"
#ifdef DISK_STATS
         "       --disk                  Show disk instead of process statistics\n"
#endif
         "       --explicit-delay        Explicitly delay between updates\n"
         "   -v, --version               Print version info\n"
         "\nArguments to long options are required for short options too.\n\n"
#else
         "   -C                   Use a monochrome color scheme\n"
         "   -d DELAY             Set the delay between updates, in tenths of seconds\n"
         "   -h                   Print this help screen\n"
         "   -s COLUMN            Sort by COLUMN (try --sort-key=help for a list)\n"
         "   -t                   Show the tree view by default\n"
         "   -u USERNAME          Show only processes of a given user\n"
         "   -p PID[,PID,PID...]  Show only the given PIDs\n"
         "   -v                   Print version info\n"
         "\n"
#endif
         "Press F1 inside htop for online help.\n"
         "See 'man htop' for more information.\n",
         name);
}

// ----------------------------------------

typedef struct CommandLineSettings_ {
   Hashtable* pidWhiteList;
   uid_t userId;
   int sortKey;
   int delay;
   bool useColors;
   bool treeView;
   bool explicit_delay;
#ifdef DISK_STATS
   bool disk;
#endif
} CommandLineSettings;

static CommandLineSettings parseArguments(int argc, char** argv) {
   const char *sort_key = NULL;
   CommandLineSettings flags = {
      .pidWhiteList = NULL,
      .userId = -1, // -1 is guaranteed to be an invalid uid_t (see setreuid(2))
      .sortKey = 0,
      .delay = -1,
      .useColors = true,
      .treeView = false,
      .explicit_delay = false,
   };

#define HTOP_LONG_OPTION_EXPLICIT_DELAY (1 << 8)
#define HTOP_LONG_OPTION_DISK (2 << 8)
#ifdef HAVE_GETOPT_LONG
   static struct option long_opts[] = {
      { "help",           no_argument,       NULL, 'h' },
      { "version",        no_argument,       NULL, 'v' },
      { "delay",          required_argument, NULL, 'd' },
      { "sort-key",       required_argument, NULL, 's' },
      { "user",           required_argument, NULL, 'u' },
      { "no-color",       no_argument,       NULL, 'C' },
      { "no-colour",      no_argument,       NULL, 'C' },
      { "tree",           no_argument,       NULL, 't' },
      { "pid",            required_argument, NULL, 'p' },
      { "explicit-delay", no_argument,       NULL, HTOP_LONG_OPTION_EXPLICIT_DELAY },
#ifdef DISK_STATS
      { "disk",           no_argument,       NULL, HTOP_LONG_OPTION_DISK },
#endif
      { NULL, 0, NULL, 0 }
   };
#endif
   /* Parse options */
   while(true) {
#ifdef HAVE_GETOPT_LONG
      int opt = getopt_long(argc, argv, "hvCs:td:u:p:", long_opts, NULL);
#else
      int opt = getopt(argc, argv, "hvCs:td:u:p:");
#endif
      if (opt == -1) break;
      switch (opt) {
            char *p;
         case HTOP_LONG_OPTION_EXPLICIT_DELAY:
            flags.explicit_delay = true;
            break;
#ifdef DISK_STATS
         case HTOP_LONG_OPTION_DISK:
            flags.disk = true;
            break;
#endif
         case 'h':
            print_usage(stdout, argv[0]);
            exit(0);
         case 'v':
            printVersionFlag();
            exit(0);
         case 's':
            sort_key = optarg;
            break;
         case 'd':
            if (sscanf(optarg, "%16d", &(flags.delay)) == 1) {
               if (flags.delay < 1) flags.delay = 1;
               if (flags.delay > 100) flags.delay = 100;
            } else {
               fprintf(stderr, "Error: invalid delay value \"%s\".\n", optarg);
               exit(-1);
            }
            break;
         case 'u':
            if (!Action_setUserOnly(optarg, &(flags.userId))) {
               fprintf(stderr, "Error: invalid user \"%s\".\n", optarg);
               exit(-1);
            }
            break;
         case 'C':
            flags.useColors = false;
            break;
         case 't':
            flags.treeView = true;
            break;
         case 'p':
            if(!flags.pidWhiteList) {
               flags.pidWhiteList = Hashtable_new(8, false);
            }
            p = optarg;
            do {
               long int pid = strtol(p, &p, 0);
               if(*p && *p != ',') {
                  fprintf(stderr, "Error: invalid PID list \"%s\".\n", optarg);
                  exit(-1);
               }
               if(pid < 0) {
                  fprintf(stderr, "Error: invalid PID %ld.\n", pid);
                  exit(-1);
               }
               Hashtable_put(flags.pidWhiteList, pid, (void *) 1);
            } while(*p++);
            break;
         default:
#ifdef __ANDROID__
            fputc('\n', stderr);
#endif
            exit(-1);
      }
   }
#ifdef DISK_STATS
   if(flags.disk) {
      if(flags.pidWhiteList) {
         fputs("Error: cannot specific PIDs for disk statistics mode\n", stderr);
         exit(-1);
      }
      if(flags.userId != (uid_t)-1) {
         fputs("Error: cannot specific user for disk statistics mode\n", stderr);
         exit(-1);
      }
      if(flags.treeView) {
         fputs("Error: disk statistics mode doesn't support tree view\n", stderr);
         exit(-1);
      }
   }
#endif
   if(sort_key) {
#ifdef DISK_STATS
      const FieldData *field_data = flags.disk ? Disk_fields : Process_fields;
      unsigned int nfields = flags.disk ? Disk_field_count : Platform_numberOfFields;
#else
      const FieldData *field_data = Process_fields;
      unsigned int nfields = Platform_numberOfFields;
#endif
      if (strcmp(sort_key, "help") == 0) {
         for (unsigned int j = 1; j < nfields; j++) {
            const char* name = field_data[j].name;
            if (name) puts(name);
         }
         exit(0);
      }
      flags.sortKey = ColumnsPanel_fieldNameToIndex(field_data, nfields, sort_key);
      if (flags.sortKey == -1) {
         fprintf(stderr, "Error: invalid column \"%s\".\n", sort_key);
         exit(-1);
      }
   }
   return flags;
}

static void millisleep(unsigned long millisec) {
#ifdef HAVE_NANOSLEEP
   struct timespec req = {
      .tv_sec = 0,
      .tv_nsec = millisec * 1000000L
   };
   while(nanosleep(&req,&req)==-1) {
      continue;
   }
#else
	unsigned int sec = millisec / 1000;
	if(sec) sleep(sec);
	millisec %= 1000;
	usleep(millisec);
#endif
}

int main(int argc, char** argv) {
   const char *lc_ctype = getenv("LC_CTYPE");
   if(!lc_ctype) {
      lc_ctype = getenv("LC_ALL");
      if(!lc_ctype) lc_ctype = "";
   }
   setlocale(LC_CTYPE, lc_ctype);

   CommandLineSettings flags = parseArguments(argc, argv); // may exit(3)

#ifdef HAVE_PROC
   if (access(PROCDIR, R_OK) < 0) {
      fprintf(stderr, "Error: could not read procfs (compiled to look in %s).\n", PROCDIR);
      return 1;
   }
#endif
   Platform_init();
   CRT_initColorSchemes();
   Process_setupColumnWidths();
   UsersTable* ut = UsersTable_new();
   ProcessList* pl = ProcessList_new(ut, flags.pidWhiteList, flags.userId);
   Settings* settings = Settings_new(pl->cpuCount, Platform_haveSwap());
   pl->settings = settings;

   Header* header = Header_new(pl, settings, 2);

   Header_populateFromSettings(header);

   if (flags.delay != -1) settings->delay = flags.delay;
   // Use built-in MONOCHROME_COLOR_SCHEME for '--no-color'
   if (!flags.useColors) settings->colorScheme = MONOCHROME_COLOR_SCHEME;
   if (flags.treeView) settings->treeView = true;
   if(flags.explicit_delay) settings->explicit_delay = true;

   CRT_init(settings);
   CRT_setMouse(settings->use_mouse);
   CRT_setExplicitDelay(settings->explicit_delay);

   Panel *panel;
#ifdef DISK_STATS
   DiskList *disk_list = NULL;
   settings->disk_mode = flags.disk;
   if(flags.disk) {
      disk_list = DiskList_new(settings);
      panel = (Panel *)DiskPanel_new(settings, disk_list, header);
      DiskList_setPanel(disk_list, panel);
      header->disk_list = disk_list;
   } else
#endif
   {
      MainPanel *process_panel = MainPanel_new();
      panel = (Panel *)process_panel;
      ProcessList_setPanel(pl, panel);
      MainPanel_updateTreeFunctions(process_panel, settings->treeView);
   }

   if (flags.sortKey > 0) {
#ifdef DISK_STATS
      if(flags.disk) settings->disk_sort_key = flags.sortKey;
      else
#endif
      {
         settings->sortKey = flags.sortKey;
         settings->treeView = false;
      }
      settings->direction = 1;
   }
#ifdef DISK_STATS
   if(flags.disk) DiskList_printHeader(disk_list, Panel_getHeader(panel));
   else
#endif
   ProcessList_printHeader(pl, Panel_getHeader(panel));

#ifdef DISK_STATS
   if(!flags.disk)
#endif
   {
      State state = {
         .settings = settings,
         .ut = ut,
         .pl = pl,
         .panel = panel,
         .header = header,
         .repeat = 1
      };
      MainPanel_setState((MainPanel *)panel, &state);
   }

   ScreenManager* scr = ScreenManager_new(0, header->height, 0, -1, HORIZONTAL, header, settings, true);
   ScreenManager_add(scr, panel, -1);

#ifdef DISK_STATS
   if(flags.disk) {
      DiskList_scan(disk_list, 0);
      ProcessList_scan(pl, true);
   } else
#endif
   ProcessList_scan(pl, false);
   millisleep(75);
#ifdef DISK_STATS
   ProcessList_scan(pl, flags.disk);
#else
   ProcessList_scan(pl, false);
#endif

   ScreenManager_run(scr, NULL, NULL);
   attron(CRT_colors[HTOP_DEFAULT_COLOR]);
   mvhline(LINES-1, 0, ' ', COLS);
   attroff(CRT_colors[HTOP_DEFAULT_COLOR]);
   refresh();
   CRT_done();
   if (settings->changed) Settings_write(settings);
   Header_delete(header);
   ProcessList_delete(pl);
#ifdef DISK_STATS
   if(flags.disk) DiskList_delete(disk_list);
#endif

   ScreenManager_delete(scr);
   UsersTable_delete(ut);
   Settings_delete(settings);
   if(flags.pidWhiteList) {
      Hashtable_delete(flags.pidWhiteList);
   }

   return 0;
}
